invenio_rdm_records/resources/urls.py

```diff
@@ -0,0 +1,57 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) 2023 Northwestern University.
+#
+# Invenio-RDM-Records is free software; you can redistribute it and/or modify
+# it under the terms of the MIT License; see LICENSE file for more details.
+
+"""Convenient URL generation.
+
+InvenioRDM poses challenges to url generation that Flask's url_for cannot handle out
+of the gate.
+
+- InvenioRDM is actually 2 applications mounted on different url_prefixes:
+  `url_for` in the API application isn't aware of the UI application endpoints
+- The endpoint names are relatively hidden / spread out and APP_RDM_ROUTES is usually
+  the interface to name them.
+- All other url generation use cases need to interact with this: generating urls outside
+  of a request context, generating canonical urls...
+
+This module contains minimal methods to generate URLs correctly without much
+engineering. Over time, it can be made more abstract, complex and powerful and even
+extracted into its own package to solve url generation across InvenioRDM once and for
+all.
+"""
+
+from flask import current_app
+
+
+def record_url_for(_app="ui", pid_value=""):
```

Comment:
Comment: I understand that the current link generation doesn't work for this use case, but I'm concerned that introducing these two methods will muddy the picture of where link generation should happen.

Would a possible solution be to return a URITemplate in the ``links`` or ``linktpls`` which could be used to generate the links you need. I.e. these two methods would take as input a URItemplate which was generated by the service layer and simply render the string.

Reply:
Hmm this is a longer discussion. The practical piece I got from the comment is to rely on the already generated links that are part of the record projection that is being serialized into `json+linkset` in the `schema.py`. I've adapted the code to rely on that e.g., https://github.com/inveniosoftware/invenio-rdm-records/pull/1404/files#diff-32235a765fef3b5ef2104531e468805d799d3d8b9db5b501d09295b4a365b4eaR33 :+1: .
The download link is not part of the provided serialized links, so I kept the use of `download_url_for` there. I also kept `record_url_for` in the header link generation, since we don't necessarily have access to those links there either.  

We can have a longer discussion about link generation in a dedicated ticket. Main desires for me are (to repeat some from the docs from urls.py): 
- avoid already potential conflicts between routes defined in `APP_RDM_ROUTES` and routes defined in `RDMRecordServiceConfig.links_item` (link generation is already muddied)
- be able to generate links anywhere with the same call pattern and different arguments i.e. like `url_for` but also from either UI or API application (in or out of request context). Maybe name it `inveniordm_url_for` 
- have well defined endpoint names to support that